psl-core/src/main/java/org/linqs/psl/reasoner/dcd/DCDReasoner.java

```diff
@@ -55,67 +56,79 @@ public double optimize(TermStore baseTermStore) {
 
+            termCount = 0;
+            objective = 0.0;
+
             for (DCDObjectiveTerm term : termStore) {
-                term.minimize(truncateEveryStep, variableValues);
+                if (oldVariableValues != null) {
+                    objective += term.evaluate(oldVariableValues);
+                }
+
+                termCount++;
+                term.minimize(truncateEveryStep, termStore.getVariableValues(), termStore.getVariableAtoms());
             }
 
             // If we are truncating every step, then the variables are already in valid state.
             if (!truncateEveryStep) {
-                for (RandomVariableAtom variable : termStore.getVariables()) {
-                    variable.setValue(Math.max(Math.min(variable.getValue(), 1.0f), 0.0f));
+                float[] variableValues = termStore.getVariableValues();
+                for (int i = 0; i < termStore.getNumVariables(); i++) {
+                    variableValues[i] = Math.max(0.0f, Math.min(1.0f, variableValues[i]));
                 }
             }
 
             long end = System.currentTimeMillis();
-
-            oldObjective = objective;
-            objective = computeObjective(termStore, variableValues);
             totalTime += end - start;
 
             if (log.isTraceEnabled()) {
-                log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}",
-                        iteration, objective, (end - start), totalTime);
+                log.trace("Iteration {} -- Objective: {}, Normalized Objective: {}, Iteration Time: {}, Total Optimization Time: {}",
+                        iteration, objective, objective / termCount, (end - start), totalTime);
             }
 
             iteration++;
             termStore.iterationComplete();
 
-            if (breakOptimization(iteration, objective, oldObjective)) {
+            if (breakOptimization(iteration, objective, oldObjective, termCount)) {
                 break;
             }
+
+            // Keep track of the old variables for a deferred objective computation.
+            if (oldVariableValues == null) {
+                oldVariableValues = Arrays.copyOf(termStore.getVariableValues(), termStore.getVariableValues().length);
+                oldObjective = Double.POSITIVE_INFINITY;
+            } else {
+                System.arraycopy(termStore.getVariableValues(), 0, oldVariableValues, 0, oldVariableValues.length);
+                oldObjective = objective;
+            }
         }
 
-        log.info("Optimization completed in {} iterations. Objective: {}, Total Optimiztion Time: {}",
-                iteration - 1, objective, totalTime);
-        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termStore.size());
+        objective = computeObjective(termStore);
+        log.info("Optimization completed in {} iterations. Objective: {}, Normalized Objective: {}, Total Optimization Time: {}",
```

Comment:
Try to align the DCD and SGD logs as much as possible (without doing any significant additional computations).